Info Encoding Draft Spec

keks opened the pull request:
> mix: context here is we're looking for a way to encode infos ... what's the encode function?
> ```js
> function MakeDeriver(encode, feedId, prevMsgId) {
>    return func(key, info, len) {
>        data = copy(info)
>        data.feedId = feedId
>        data.prevMsgId = prevMsgId
>        return HKDF.Expand(key, encode(data), len)
>    }
> }
> 
> Derive = MakeDeriver(standardEncode, feedId, prevMsgId)
> ```
>

I first sketched
- the "flat" length-prefixed encoding (Shallow Length Prefix (SLP) encoding)
- nested length-prefixed encoding without type hints (is this a buffer or a list) (Recursive Length Prefix (RLP) encoding)
- nested length-prefixed encoding with type hints

I realized that since everyone who does the derivation _knows_ which derivations they are supposed to do, there is a sort of (up until now) implicit _schema_.

After thinking about it for a while I could not come up with a reason why whoever performs the derivations would not have such a schema. For our base case, the schema is defined in the core box2 spec. For extensions, it will be in the respective extension's spec.

Therefore, I realized that "type confusion" is not possible, so I scratched the version with type hints.

Since the nested/recursive encoding is just a special case of the flat one, where the buffer contents has a special meaning (i.e. "is of a special type"), I realized that this is safe to use **but currently we don't need it**.

So I propose using the "flat" SLP for now, and use RLP is extensions where we really need it. After all, it only changes the contents of the buffers in the list, not the encoding of the outer list itself.

Open questions:
- In the Requirements section there is a clunky sentence that needs to be rephrased a bit.
- I just guessed big endian. I am not aware of any real pro/con for either. Any ideas/opinions?
- I only now see that there are some rendering issues, e.g. the `<a>` tag for the code heading didn't work. I'll fix that later.
- I'd be happy if you could review this! I'm sure it's still pretty rough.

keks committed:
add info encoding draft spec

mixmix commented:
I modified this to add context at the start of your message

mixmix commented on info_encoding.md:
```diff
@@ -0,0 +1,143 @@
+This is enough for most simple operations, and can be extended to
+key-value stores.
+
+To get a simpler encoding, we do not require schemalessness.
```
is this a double negative. Could we just say 

> To get a simpler encoding, we define a schema for encoding the info field

keks replied:
Hm, to me schemalessness is a special word... How about "we require the user to adhere to a schema"?

keks replied:
Hm, that makes the sentence afterwards weird.

The thing is that we don't want to define the schema ourselves, we want to user to know what they are doing.

keks replied:
How about:
> A property of many structured data encodings is schemalessness. To get a simpler encoding, we do not require this property. Therefore, derivation steps need to define a schema for the encoding of the info field.

mixmix replied:
It's just an odd structure to me. Can we go : 

> A property of many structured data encodings is schemalessness. 
> To get a simpler encoding, we are instead requiring a schema. 
> The definition of a schema for each encoding (i.e. what's in the `list` of info, and what the order is) is therefore required. 
> In this module, see `DeriveSecret` which is using this SLP

mixmix commented on info_encoding.md:
```diff
@@ -0,0 +1,143 @@
+Therefore, derivation steps need to define a schema for the encoding
+of the info field.
+For the core derivations, this is given.
+Custom applications need to provide this as well.
```
Custom applications (defined in `extensions) need to provide this as well.

mixmix commented on info_encoding.md:
```diff
@@ -0,0 +1,143 @@
+
+Specifically, the encoding is the element-wise concatenation of the the
+length of the elements and the element.
+The length is encoded as big endian uint16. (REALLY BIG ENDIAN? PRO/CON?)
```
I know we used  little-endian encoding for the offset in the header: 
```js 
header.writeUInt16LE(offset, 0)
```

I don't know the pros and cons of each.
Could this be an example of "lets use a consistent pattern to make implementing the spec easier"?

keks replied:
If we'd try to sort the data by bytes the endianness would be important, but yes, I think in this case consistency is key. Maybe we should define a byte order somewhere and refer to that, so we can make up our minds later and enforce consistency.

mixmix replied:
Yep, but there's only 2 instances at the moment. Probably too early to abstract

mixmix replied:
TOO LATE!

mixmix commented on info_encoding.md:
````diff
@@ -0,0 +1,143 @@
+```
+is encoded as:
+```
+uint16BE(len(a)) || a || uint16BE(len(b)) || b || uint16BE(len(c)) || c
````
is `length-prefixed` the same as scuttlebutt's `offset` in the literature?

oh, no I see it's what I expect from the name

mixmix commented on info_encoding.md:
````diff
@@ -0,0 +1,143 @@
+## Code
+
+```
+function encode(l List) Buffer {
````
let's use a variable that doesn't look like `1`, `list` is probably fine

mixmix replied:
Maybe this for the JS example

---
```js
function encode (list) {
  var out = Buffer.alloc(0)

  for (let buf in list) {
    const length = Buffer.alloc(2)
    length.writeUInt16BE(buf.length, 0)

    out = Buffer.concat([out, length, buf])
  }

  return out
}
```

where 
- `list` is an Array of Buffers
- `encode` returns a Buffer

mixmix commented:
@keks this read fairly clearly to me:

:heavy_check_mark:  I'm confident I'd know how to implement this
:x: I don't have the cryptography context to fully review this
:question: I think it could benefit from an example with actual infos

e.g. 

> to encode the infos for a secret the schema might be:
> ``` 
> ( feed_id, previous_msg_id, label )
> ```
> where each of these are buffers, the "label" being a "utf8" encoded string in the form of a Buffer

keks committed:
Review mix and more

- removed newlines for prettier rendering
- better wording around double negation
- clarified where custom applications specify their own schemas
- removed all references to endiannes except in Notations and
  Definitions section, and in code
- don't use l as a variable because it often looks like 1 or I
- add section on how to use lists for key-value lists
- add examples and more explanations around features they use

keks commented:
Hey @mixmix, thanks for the review! There are two items I didn't want to resolve without checking back with you, would be cool if you could confirm. Also I added a lot of new stuff, could you review that as well?

mixmix commented on info_encoding.md:
````diff
@@ -1,116 +1,108 @@
+	assert i >= 0
+	return uint16(i).encode()
+}
+```
````
This pseudocode confuses me, because it's too far from languages I'm familiar with. (i.e. I might not be the best to review this).
Are you just defining `UInt16LE` (unsigned integer, 16 bit, little-endian)?
Isn't this fairly common?

mixmix replied:
Also, you say `method ()` and `function ()`

mixmix replied:
I might be peculiar, but I personally prefer a "pictionary style" approach to introducing new ideas.

1. start fuzzy, big picture
  >  "we store things as a concatenated list, so order matters"
2. add more detail
  > "we prefix each thing we concatenate with it's length!"
3. get _real_ specific
  > the encoding of that length is `UInt16LE` , here's exactly what that means, and here's what it all looks like in details when you encode this example string with this example "schema" (the definition of what's in the list and the order)

_This is just a style-comment. It's a style I find useful for teaching and introducing new ideas (I was a teacher for quite a few years). We don't have to change anything, this is pretty clear. This plus test-vectors will lock it all in_

mixmix commented on info_encoding.md:
```diff
@@ -1,116 +1,108 @@
-This is enough for most simple operations, and can be extended to
-key-value stores.
+We require the format to support lists of buffers. This is enough for most simple operations, and can be extended to key-value stores.
+A property of many structured data encodings is schemalessness. To get a simpler encoding, we do not require this property. Therefore, derivation steps need to define a schema for the encoding of the info field. For the core derivations, this is given. Custom applications that are defined in extensions need to provide this as well.
```
This is clear enough.
I think let's just be specific about the core examples ... i.e. "see DeriveSecret"

mixmix commented on info_encoding.md:
```diff
@@ -1,116 +1,108 @@
-is supposed to be used for.
-If they are not, the adversary got them to execute their code already,
-which means that particular user is in trouble anyways (REPHRASE).
+ Note that schemalessness would not help in this context, because all users who derive a key agree, by following the specification, on what that key is supposed to be used for up front, so even in a schema
```
incomplete sentence

mixmix commented on info_encoding.md:
```diff
@@ -1,116 +1,108 @@
-The proposed format is called _shallow length prefixed (SLP)_ encoding.
-Shallow, because it does not specify nesting.
-Length-prefixed, because each element is prefixed with its length.
+The proposed format is called _shallow length prefixed (SLP)_ encoding. Shallow, because it does not specify nesting. Length-prefixed, because each element is prefixed with its length.
```
it's not a **proposed** format, this isn't a scientific paper, we're actually building things.

i.e. remove "proposed"

mixmix commented on info_encoding.md:
```diff
@@ -1,116 +1,108 @@
-length of the elements and the element.
-The length is encoded as big endian uint16. (REALLY BIG ENDIAN? PRO/CON?)
-The pseudocode of the algorithm can be found in section [Code].
+Specifically, the encoding is the element-wise concatenation of the the length of the elements and the element. The length is a uint16 encoded using common byte order. The pseudocode of the algorithm can be found in section [Code].
```
:fire: Markdown links like this are (I think) `[the Code section][Code]`

mixmix commented on info_encoding.md:
````diff
@@ -119,25 +111,89 @@ function encode(l List) Buffer {
+
+```js
+function binEncodeUInt16(target, number) {
+    target.writeUInt16BE(number, 0)
````
:fire:  This should be `LE` not `BE`

mixmix commented on info_encoding.md:
````diff
@@ -119,25 +111,89 @@ function encode(l List) Buffer {
+```
+This would be list-encoded by alternating between keys and values:
+```
+("purpose", "box2", "type", "read key", "feed", "@feedID", "prev", "%msgID")
````
:fire:  I think you need to say explicitly:

> here the "schema" is the expected keys and their values, along with **the order in which they're encoded**

mixmix replied:
:fire:  You haven't talked here about how these strings were converted to buffers. Should I know how to do this?

mixmix commented on info_encoding.md:
```diff
@@ -119,25 +111,89 @@ function encode(l List) Buffer {
+     b o x 2       r e a d   k e y       @ f e e d I D       @ m s g I D
+0400 626f7832 0800 72656164206b6579 0700 40666565644944 0600 406d73674944
+^^^^          ^^^^                  ^^^^                ^^^^
+length        length                length              length
```
This simpler example is much easier to read, maybe just have this

mixmix commented on info_encoding.md:
````diff
@@ -1,116 +1,108 @@
+```
+is given by
+```
+("key1", "value1", "key2", "value2")
````
replace this example here with the sweet example below (with just 2 keys.)

mixmix commented on info_encoding.md:
````diff
@@ -1,116 +1,108 @@
 ```
 
 ## Nested Data
 
````
:fire:  This whole "Nested Data" section needs to be right at the end.
We're not using it yet, we 100% don't want people to get confused by this as opposed to the thing they need right now.

I'm glad you wrote the ideas down though, they'll be useful in future perhaps

mixmix commented on info_encoding.md:
````diff
@@ -1,116 +1,108 @@
 ```
 
-<a href name=Code>&nsbp;</a>
+<a name="section-code"> </a>
````
oh, you don't need this for many markdowns (e.g. gh-markdown).
Titles get automatically converted into kebab-cased anchors.

mixmix commented:
# Summary of review

@keks this is looking a lot clearer to me.
I'll put a :fire: by the things which I think are _really important_ to change before merge.

I think once we've got this encoding we just need to add a clear spec about how we use it in DeriveSecret and we're ready to go!

---

Really appreciate the detail you've put into this Keks.
Great work   (ﾉ´ヮ´)ﾉ*:･ﾟ✧

mixmix committed:
Merge branch 'first_pass' of github.com:ssbc/box2-spec into info_encoding

mixmix committed:
mix's changes to slp_encoding

mixmix commented:
@keks I made several small changes I'd recommended, and have linked this in to README.md.

You can see the changes I made in the commit history above. Hope that's all good :)
